clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -202,7 +205,23 @@ public String toString() {
          * Mark that we want data from this partition in the upcoming fetch.
          */
         public void add(TopicPartition topicPartition, PartitionData data) {
+            if (sessionPartitions.containsKey(topicPartition)) {
```

Comment:
I think you may be able to `.put` the added PartitionData into sessionPartitions, and capture the value returned by the `.put`, as it will return the previous value. If the previous value was null, then we add to added, if it was not null we add it to altered. This will save us the additional containsKey check.

Reply:
Sounds good.

Reply:
If we didn't care about maintaining the altered and added lists (which are only used for debug level logging...), I would suggest that we simply add to next and then perform a putAll to sessionPartitions on build. It would likely perform better if we are OK with losing a little info at debug levels.